### PR TITLE
Non static state

### PR DESCRIPTION
**Pull request by luleyleo**
Still needs some polishing, but it's mostly done.

### CODE REVIEW

**Comment by Saethox** [comment]
I had some trouble building this under Windows (my fault for using it :D), which was mostly because of `coco-sys`.

Some notes:
- The python executable under Windows is usually called `python` (not `python3`). Maybe use something like this in the build script?

```rust
#[cfg(target_os = "linux")]
let python_path = "python3";
#[cfg(target_os = "windows")]
let python_path = "python";

Command::new(python_path)
    .arg(build_script_path)
    .arg("prepare-build-rust")
    .output()
    .expect("failed to run prepare-build-rust command");
```

- `libclang` can be installed under Windows using the [choco llvm package](https://community.chocolatey.org/packages/llvm) (Maybe add this to the README of `coco-rs`?)
```ps
choco install llvm
```

With those, it compiles properly.

File: src/framework/configuration.rs [hunk 1]
```diff
@@ -202,7 +202,7 @@ impl<P: Problem> ConfigurationBuilder<P> {
     }
 
     pub fn evaluate_sequential(self) -> Self {
```
Comment (Saethox) [inline src/framework/configuration.rs:204]: This could be renamed to `evaluate`.

**Review by Saethox — APPROVED** [review]
Looks good to me in general. One thing that's missing is to adapt the remaining heuristic definitions to the new logging.

The `Logger` should also probably insert an empty `LogSet` when none is present, the `bmf.rs` and `tsp.rs` examples currently panic because no `LogSet` is present by default.

The `coco.rs` example also currently panics because of some internal logging problem.

**Review by HeleNoir — APPROVED** [review]

File: src/state/container/mod.rs [hunk 2]
```diff
@@ -58,43 +58,82 @@ impl State {
...
     }
 
+    /// Tires to find an inner map containing T.
+    fn find<T: CustomState<'a>>(&self) -> Option<&Self> {
+        if self.map.has::<T>() {
+            Some(self)
+        } else {
+            self.parent()
+        }
+    }
```
Comment (Saethox) [inline src/state/container/mod.rs:72]: Shouldn't this be `self.parent.find::<T>`? The same for `find_mut()`.

File: src/state/container/mod.rs [hunk 3]
```diff
@@ -58,43 +58,82 @@ impl State {
...
+    #[track_caller]
+    pub fn holding<T: CustomState<'a>>(&mut self, code: impl FnOnce(&mut T, &mut Self)) {
+        let state_with_t = self.find_mut::<T>().unwrap() as *mut Self;
+
+        let mut instance = self.take::<T>();
+        code(&mut instance, self);
+
+        // Insert the state where it was before.
+        // This is save, because inner states can not be removed.
+        unsafe { state_with_t.as_mut().unwrap().insert(instance) };
```
Comment (Saethox) [inline src/state/container/mod.rs:123]: What exactly was the issue here that we need `unsafe`?

Comment (luleyleo) [reply src/state/container/mod.rs]: I've changed it to a solution without `unsafe`.

**Comment by luleyleo** [comment]
Everything should be working now.

**Comment by Saethox** [comment]
LGTM, I think we can merge.

**Pull request merged**